base/utfcheck.jl

```diff
@@ -0,0 +1,255 @@
+                # check for surrogate pairs, make sure correct
+                if is_surrogate_codepoint(ch)
+                    !is_surrogate_lead(ch) && utf_errfunc(UTF_ERR_NOT_LEAD, pos-2, ch)
+                    # next character *must* be a trailing surrogate character
+                    (pos + 3 > len) && utf_errfunc(UTF_ERR_MISSING_SURROGATE, pos-2, ch)
+                    byt = dat[pos += 1]
+                    (byt != 0xed) && utf_errfunc(UTF_ERR_NOT_TRAIL, pos, byt)
+                    surr = get_continuation(0x0000d, dat[pos += 1], pos)
+                    surr = get_continuation(surr, dat[pos += 1], pos)
+                    !is_surrogate_trail(surr) && utf_errfunc(UTF_ERR_NOT_TRAIL, pos-2, surr)
+                    (options & UTF_NO_SURROGATES) != 0 && utf_errfunc(UTF_ERR_SURROGATE, pos-2, surr)
+                    flags |= UTF_SURROGATE
+                    num4byte += 1
+                elseif ch > 0x07ff
+                    num3byte += 1
+                elseif (options & UTF_ACCEPT_LONG) != 0
+                    flags |= UTF_LONG
+                    num2byte += 1
+                else
+                    utf_errfunc(UTF_ERR_LONG, pos-2, ch)
+                end
+            elseif ch < 0xf5
+                # 4-byte UTF-8 sequence (i.e. characters > 0xffff)
+                (pos + 3 > len) && utf_errfunc(UTF_ERR_SHORT, pos, ch)
+                ch = get_continuation(ch & 0x07, dat[pos += 1], pos)
+                ch = get_continuation(ch, dat[pos += 1], pos)
+                ch = get_continuation(ch, dat[pos += 1], pos)
+                if ch > 0x10ffff
+                    utf_errfunc(UTF_ERR_INVALID, pos-3, ch)
+                elseif ch > 0xffff
+                    num4byte += 1
+                elseif is_surrogate_codepoint(ch)
+                    utf_errfunc(UTF_ERR_SURROGATE, pos-3, ch)
+                elseif (options & UTF_ACCEPT_LONG) != 0
+                    # This is an overly long encode character
+                    flags |= UTF_LONG
+                    if ch > 0x7ff
+                        num3byte += 1
+                    elseif ch > 0x7f
+                        num2byte += 1
+                    end
+                else
+                    utf_errfunc(UTF_ERR_LONG, pos-2, ch)
+                end
+            else
+                utf_errfunc(UTF_ERR_INVALID, pos, ch)
+            end
+        end
+    end
+    num3byte != 0 && (flags |= UTF_UNICODE3)
+    num4byte != 0 && (flags |= UTF_UNICODE4)
+    return totalchar, flags, num4byte, num3byte, num2byte
+end
+
+#=
+@doc """
+@brief      Validates and calculates number of characters in a UTF-16 encoded vector of UInt16
+
+@param[in]  dat     Vector{UInt16}
+@param[in]  options flags to determine error handling (default 0)
+
+@return     (total characters, flags, 4-byte, 3-byte, 2-byte)
+@throws     ArgumentError
+""" ->
+=#
+function check_string_utf16(dat::Vector{UInt16}, len::Int)
```

Comment:
I have no problem with people learning as they go. We're not punishing you because you didn't do it the proscribed way, "because I said so". This PR hasn't been merged because it's nearly impossible to effectively review and continues to have major issues. For the amount of effort that's been spent on this PR by you – not to mention everyone else – you could easily have gone back and broken it into smaller changes that are actually reviewable and which can be merged one at a time without so much risk. One obvious factoring is bug fixes versus performance improvements. Yet still, here we are wasting time on an unreviewable pull request.